docs/field_presence.md

```diff
@@ -0,0 +1,459 @@
+# Application note: Field presence
+
+This application note explains the various presence tracking disciplies for protobuf fields. It also explains how to enable experimental support for explicit presence tracking for singular proto3 fields with basic types.
+
+## Background
+
+_Field presence_ is the notion of whether a protobuf field has a value. Different types of fields follow different _presence disciplines_, or semantics for translating between the _API representation_ and the _serialized representation_.
+
+The _implicit_ presence discipline relies upon the field value itself to express notional presence. This can also be called _in-band_ presence tracking, since presence (or, usually, absence) is represented by an otherwise valid value.
+
+The _explicit_ presence discipline relies upon state tracking which is orthogonal to a field value to express notional presence. This could also be called _external_ or _out-of-band_ presence tracking.
+
+### Presence in _tag-value stream_ (wire format) serialization
+
+The wire format is a stream of tagged, _self-delimiting_ values. By definition, the wire format represents a sequence of _present_ values. In other words, every value found within a serialization represents a _present_ field; furthermore, the serialization contains no information about not-present values.
+
+The generated API for a proto message includes (de)serialization definitions which translate between API types and a stream of definitionally _present_ (tag, value) pairs. This translation is designed to be forward- and backward-compatibile across changes to the message definition; however, this compatibility introduces some (perhaps surprising) considerations when deserializing wire-formatted messages:
+
+-   "Empty" length-delimited values (such as empty strings or empty message-typed fields) can be represented in serialized values. However, if the generated API does not track presence, then these values may not be re-serialized; i.e., the empty field may be "not present" after a serialization round-trip.
+-   When deserializing, duplicate field values may be handled in different ways.
+	-   Duplicate `repeated` fields are typically appended to the field's API representation. (Note that serializing a _packed_ repeated field produces only one, length-delimited value in the tag stream.)
+	-   Duplicate `optional` field values follow the rule that "the last one wins."
+-   `oneof` fields expose the API-level invariant that only one field is set at a time. However, the wire format may include multiple (tag, value) pairs which notionally belong to the `oneof`. Similar to `optional` fields, the generated API follows the "last one wins" rule.
+-   Out-of-range values are not returned for fields in the generated API. However, out-of-range values may be stored as _unknown fields_ in the API, even though the wire-format tag is recognized.
```

Comment:
I think this is a bit misleading. Only enums store out-of-range values in the unknown field set when the wire format matches. For every other field type (integers and bool), out of range values are truncated and returned through generated accessors.

Reply:
Good catch... updated.